capabilities documentation (#296)
oasis-open/odata-rapid: docs/rsdl/rapid-rsdl-capabilities.md

Without further constraints this would allow a huge number of URLs that a service would need to support. And it is important not just to specify which paths are allowed, but also to specify different functionality and behaviors supported for different paths.

The following sections introduce the notion of paths and capabilities that the service provides per path.

## Example service

For the examples in the section of this document, we assume the following type definitions. This is a simplified service that provides a top level orders collection together with their (contained) order items. Each item references a SKU, that can be accessed via the top level entity set. 

```rsdl
type Order {
    key id: String
    created: DateTime
    status: OrderStatus
    items: [OrderItem]
}

enum OrderStatus {
    Open
    Archived
    Canceled 
}

type OrderItem {
    key id: String

>     sku: *SKU

Reviewer: ```suggestion
    sku: SKU
```

We decided against using the `*` prefix for non-containment

    amount: Integer
}

type SKU {
    key id: String
    name: String
    description: String
    unitPrice: Decimal
}

service {
    orders: [Order]
    skus: [SKU]
}
```

The amount of potential URLs that the service may support is large - in larger services potentially unbounded. Just to name a few of the more important ones:

``` http
/skus, /skus/{id}, /orders, /orders/{id}, /orders/{id}/items, /orders/{id}/items/{id},
/orders/{id}/items/{id}/skus, /orders/{id}/items/{id}/skus/{id}, ...
```

The path-centric view in RSDL allows to enumerate the allowed requests and specify which  query options are supported for these requests.

## HTTP capabilities

> The first level of the above mentioned capabilities is to specify the path of theURL and the HTTP method, Here is a partial RSDL to demonstrate this.

Reviewer: ```suggestion
The first level of the above mentioned capabilities is to specify the path of the URL and the HTTP method, Here is a partial RSDL to demonstrate this.
```

``` rsdl
path /orders {
    GET { ... }
    POST { ... }
}

path /orders/{id} {
    GET { ... }
    PATCH { ... }
    DELETE { ... }
}

path /orders/{id}/items {
    GET { ... }
    POST { ... }
}

path /orders/{id}/items/{id} {
    GET { ... }
    DELETE { ... }
}

path /skus {
    GET { ... }
}

```

> The effect of this is that for each declared `path` declaration and HTTP method the service is expected to return data, data in the form as specified by the corresponding type. The service is free to respond with success messages for other combinations but the service guarantees to work for the ones listed.

Reviewer: ```suggestion
The effect of this is that for each declared `path` and HTTP method the service is expected to return data, data in the form as specified by the corresponding type. The service is free to respond with success messages for other combinations but the service guarantees to work for the ones listed.
```

> Important to notice is that the path after the `path` keyword are not literal paths but sort of templates. Many of the interesting paths in a REST service have path segments that are entity ids. this is indicated via the `{id}` syntax. the `id` in there is exactly the name of the key.

Reviewer: ```suggestion
Important to notice is that the paths after the `path` keyword are not literal paths but sort of templates. Many of the interesting paths in a REST service have path segments that are entity ids. This is indicated via the `{id}` syntax: the `id` in there is exactly the name of the key property.
```

The specific capabilities that can be used in the HTTP capabilities section instead of the `...` in the example above can vary by HTTP method. Here is an overview 

|        |        filter        |        expand        |        paging        |        count         |
|--------|:--------------------:|:--------------------:|:--------------------:|:--------------------:|
| GET    |       &#x2713;       |       &#x2713;       |       &#x2713;       |       &#x2713;       |
| POST   | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> |
| PATCH  | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> |
| PUT    | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> | &#x2713;<sup>1</sup> |
| DELETE | |                      |                      |                      |

[1] to shape the POST/PATCH/PUT response. Rarely used but supported<br/>
[2] deleting multiple items. Rarely used but supported<br/>
[!!TODO: check if filter segment on delete is allowed in RAPID ]

## Individual Query capabilities

### Filter capabilities

> The filter capability allows to specify which property can be used in which filter expression. There is a vast amount of possible filter expressions (see [odata abnf](https://github.com/oasis-tcs/odata-abnf/blob/main/abnf/odata-abnf-construction-rules.txt#L502)). Therefore, the filter capabilities allow to specify a few well-known but restrictive expressions or allow any expression.

Reviewer: ```suggestion
The filter capability allows to specify which property can be used in which filter expression. There is a vast amount of possible filter expressions (see [OData System Query Option `$filter`](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionfilter)). Therefore, the filter capabilities allow to specify a few well-known but restrictive expressions or allow any expression.
```

| operator group | comment                                                                                                               |
|----------------|-----------------------------------------------------------------------------------------------------------------------|
| eq             | `<property> eq <literal>` or <br/> `<property> in (<literal1>, <literal2>, ... ) `                                    |
| range          | `<literal> le <property> and <property> le <literal>` <br/> or equivalent expressions with operators `ge`, `gt`, `lt` |
| ranges         | a disjunction of `range` expressions                                                                                  |
| prefix         | `startswith(<property>, <literal>)`                                                                                   |
| text           | `<string op>(<property>, <literal>)`,<br/>where `<string op>` is one of `startswith`, `endswith`, `contains`         |
| any            | any expression including expressions combined with `not`, `and`, and `or`                                             |

> In RSDL this 

Reviewer: Sentence looks incomplete

```rsdl
path /orders {
    GET { 
        filter { 
            eq     { id name description createdDate fulfillmentDate }
            ranges { createdDate description }
            prefix { name }
            text   { description }                   
        }

        filter { 
            eq     except { description }                 
            ranges { createdDate description }
            prefix { name }
            text   { description }                   

>         }

Reviewer: Really two occurrences of `filter {...}`? Or should these be two separate examples?

    }
}
```

### Expand capabilities

> The expand capability allows to specify which property can be used in `$expand` query parameter. 

Reviewer: ```suggestion
The expand capability allows to specify which properties can be used in `$expand` query parameter. 
```

The count capability allows to specify if a request returning a collection can specify $count query parameter.

```rsdl
path /orders {
    GET { 
        count
    }
}
```

## Extensibility

The RSDL API capabilities allow to specify custom capabilities, so called traits, for the service to implement that are outside the scope of the patterns known by the RSDL specification. 

One example for these could be the handling of requests when throttling is needed, where the service responds with certain headers for requests that cross a throttling threshold. There are multiple established pattern (protocol) to implement this but it is not covered in RSDL directly. But it can be specified in RSDL as an extension and the service implementation can read it and configure the right behavior based on the traits in RSDL.

The individual trait is treated as simple opaque (meta-) data that is passed on the service in the form of a trait name and a list of key-value pairs. The implementation is then free to interpret the presence of the trait and the parameters as it sees fit.

```rsdl
path /orders {
    GET { 
        traits { 
            longRunningOperation
            topWithoutSkip

>             throttling { level: "premium\u12a4" }

Reviewer: Are the values always strings, or can they also be arrays, objects, numbers, booleans, i.e. "embedded JSON"?

        }
    }
}
```

## Other path based characteristics

```rsdl
path /user/{id}/recentOrders {
    # references an Order contained in /orders
    targets: /orders
    targets: unbound

>     # 'targets' ':' ( path / 'unbound' )

Reviewer: `targets` sounds like plural: better use `target`?

    GET {  }
}

path /orders {
    # top level entity set
    GET {  }

}

path /orders/{id}/items {
    # containment navigation (informally an entity set)
    GET {  }

}
```

## Syntax

!! NOTE: incomplete

``` abnf
path-capability       = "path" path "{" 
                            [ get-capabilities ]
                            [ post-capabilities ]
                            [ patch-capabilities ]
                            [ delete-capabilities ]
                        "}"

query-capabilities    = [ expand-capabilities ] / 
                        [ filter-capabilities ] / [ select-capabilities ] /

>                         [ paging-capabilities ] / [ count-capabilities ] /

Reviewer: Better move after `delete-capabilities` for "hierarchical reading".

get-capabilities      = "GET" "{" [ query-capabilities ] "}"

post-capabilities     = "POST" "{" "}"

patch-capabilities    = "PATCH" "{" "}"

> delete-capabilities   = "DELETE" "{" [ filter-capabilities ] "}"

Reviewer: ```suggestion
delete-capabilities   = "DELETE" "{" "}"
```
We decided that filter segments are too fancy
